Relayer v2 ADR and documentation

soareschen opened the pull request:
## Description

This is an ongoing work on adding documentation to #2478, as well as writing an ADR for it.

The `cargo doc` documentation for this PR is available at https://ibc-rs.netlify.app/ibc_relayer_framework/.

[Rendered ADR](https://ibc-rs.netlify.app/ibc_relayer_cosmos/docs/adr/index.html)

[Concurrency Architectures](https://ibc-rs.netlify.app/ibc_relayer_framework/docs/concurrency_architectures/index.html) ([Excalidraw Diagrams](https://app.excalidraw.com/l/4XqkU6POmGI/9oNGlMca2wF))

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

seanchen1991 commented on docs/architecture/adr-012-ibc-relayer-v2.md:
```diff
@@ -0,0 +1,744 @@
+possible states using model checking tools like TLA+ and Apalache. On the other
+hand, since the relayer framework itself is fully abstract, it is also possble
+to treat the relayer framework as a model. This can be potentially done by using
+model checking tools for Rust, such as Kani and Prusti. If that is possible,
```
Let's add links to Kani and Prusti.

soareschen replied:
Done

seanchen1991 commented on docs/architecture/adr-012-ibc-relayer-v2.md:
```diff
@@ -0,0 +1,744 @@
+Because of this, a lot of rigorous testing are required to ensure that the
+combined logic of retrying to send packets and transactions is sound. A good
+way to test that is to build a model of the concurrent system and test all
+possible states using model checking tools like TLA+ and Apalache. On the other
```
Let's add links to TLA+ and Apalache.

soareschen replied:
Done

seanchen1991 commented on docs/architecture/adr-012-ibc-relayer-v2.md:
```diff
@@ -0,0 +1,744 @@
+It is worth noting that even though the relayer components defined using
+context-generic programming are fully modular, the all-in-one traits are
+designed to be _rigid_ and _specialized_. Observant readers may notice that
+the all-in-one traits are similar to the traditional god traits that are
```
"Traditional god traits" seems like a typo, but I have no idea what's it supposed to be instead.

soareschen replied:
They are just the traits that contain too many methods, with developers keep adding new methods to it whenever they need something new. I have changed the term to monolithic traits and explain them a little bit.

seanchen1991 commented on docs/architecture/adr-012-ibc-relayer-v2.md:
```diff
@@ -0,0 +1,744 @@
+the all-in-one traits are similar to the traditional god traits that are
+commonly found in Rust code, such as `ChainHandle` in relayer v1. At the expense
+of user convenience, the all-in-one traits suffer the same limitations as other
+god traits of being very complex and inflexible. However, the main difference is
```
"god traits" shows up again here. Is this just a term I'm not familiar with?

seanchen1991 commented on docs/architecture/adr-012-ibc-relayer-v2.md:
```diff
@@ -0,0 +1,744 @@
+
+For demonstration purpose, the above code is slightly simplified from the actual
+Cosmos chain implementation. Readers are encouraged to refer to the
+`ibc-relayer-cosmos` itself to see the full implementation details.
```
Let's link to the actual `impl OfaBaseChain for MinCosmosChainContext` code here.

seanchen1991 commented on docs/architecture/adr-012-ibc-relayer-v2.md:
```diff
@@ -0,0 +1,744 @@
+
+We would also do the same implementation for the concrete relay contexts, such
+as implementing `OfaBaseRelay` for  `MinCosmosRelayContext`, which would contain
+two `MinCosmosChainContext`s. Once the traits are implemented, the relayer
```
It's not clear why there would be two `MinCosmosChainContext`s.

soareschen replied:
Made clarification that one is for the source chain, and the other is for the destination chain.

soareschen commented on docs/architecture/adr-012-ibc-relayer-v2.md:
````diff
@@ -0,0 +1,670 @@
+}
+```
+
+Readers are encouraged to refer to the documentation and source code for
````
The tricky part for this is that the documentation has not been published to [docs.rs](https://docs.rs/), so there is no official link that we can refer to.

To workaround this, I aliased the ADR inside the cargo doc for `ibc-relayer-cosmos`, so that it can be rendered as a Rustdoc page together with the rest of the documentation. I also can't put it inside `ibc-relayer-framework`, because then the documentation wouldn't be able to link to `ibc-relayer-cosmos` or the relayer constructs in `ibc-relayer`.

soareschen replied:
Here is the new rendered ADR as cargo doc: https://ibc-rs.netlify.app/ibc_relayer_cosmos/docs/adr/index.html

ljoss17 commented on docs/architecture/adr-012-ibc-relayer-v2.md:
```diff
@@ -0,0 +1,829 @@
+relayer implementation, the relay context appears to be exclusively owned by the
+packet relayer, and it is not aware of other concurrent tasks running.
+
+The relayer framework uses multiple layers of optimizations to improve the
+efficiency of relaying IBC packets. The first layer performs message
+batching per relay context, by collecting messages being sent over a relay
+context within a time frame and sending them all as a single batch if it does not
+exceed the batch size limit. This layer does the batching per relay context,
```
I am not sure to understand what the relayer framework and the optimisation layers referred here are. At the end it is specified that these optimisation layers are optional and new optimisations can easily be added. Does this mean that the three optimisation layers are examples of what can be done ?

soareschen replied:
Yes, the optimizations listed are the default that is provided, but advance users can opt out of the default if they have other options. I have clarified the given paragraph.

soareschen commented:
I have added a new documentation page for the [Concurrency Architectures](https://ibc-rs.netlify.app/ibc_relayer_framework/docs/concurrency_architectures/index.html), together with the [Excalidraw Diagrams](https://app.excalidraw.com/l/4XqkU6POmGI/9oNGlMca2wF).

I originally planned to put the concurrency architectures as a section in the ADR, but it gets too large, so I moved it to a separate document.

soareschen commented on docs/architecture/adr-012-ibc-relayer-v2.md:
```diff
@@ -0,0 +1,744 @@
+logic for the packet relayers themselves. From the perspective of the packet
+relayer implementation, the relay context appears to be exclusively owned by the
+packet relayer, and it is not aware of other concurrent tasks running.
+
```
Added new heading.

soareschen commented on docs/architecture/adr-012-ibc-relayer-v2.md:
````diff
@@ -0,0 +1,670 @@
+{ /* ... */ }
+```
+
+The `SkipUpdateClient` component is a _middleware_ component that wraps around
````
Added clarification on what middleware means here.

soareschen commented:
Let's merge this first and continue the documentation work on a follow up PR, so that we get less merge conflicts.
